### ISSUE TITLE
Download remote images and cache them to be able to use relative paths

### ISSUE THREAD

**Issue by ArcaneTSGK** [gatsbyjs/gatsby#13830]
## Summary
We are attempting to use a third party headless CMS to manage images for our Gatsby site. We want to be able to transform the absolute path 'https://managedcontent.brandname.com/imagename.jpg' to be cached in Gatsby so that we can use relative paths in our components and css.

## Relevant information
The CMS exposes GraphQL endpoints to query all of the data for which we use the plugin 'gatsby-source-graphql' to retrieve the data.

An example object that will be returned is as follows

```
promotion { 
  url: "https://managedcontent.brandname.com/imagename.jpg",
  backgroundImageUrl: "/imagename.jpg"
}
```

In the above example, we are providing the absolute path of the image where Gatsby can find it, but the path we want to use as a css property of one of the components (using emotion) is the backgroundImageUrl 

Is there any way we can achieve this?

**Comment by sidharthachatterjee** [comment]
You should be able to use the `createRemoteFileNode` helper from `gatsby-source-filesystem` (https://www.gatsbyjs.org/packages/gatsby-source-filesystem/#createremotefilenode) to manually create `File` nodes from the urls. This will download them and make them accessible from the local filesystem enabling you to do whatever you want with them:

- like the css background thing you described
- optimisations with `gatsby-plugin-sharp` 😉 

Paging @pieh in case there is a "better" way to do this!

**Comment by muescha** [comment]
Which CMS you use?

**Comment by jonniebigodes** [comment]
@ArcaneTSGK based on the information at hand you can create your own [plugin](https://www.gatsbyjs.org/docs/pixabay-source-plugin-tutorial/) to fetch the content and then as @sidharthachatterjee said make use of `createRemoteFileNode`. Alternatively you can use the lifecycle apis, more specifically `onPreBoostrap` and fetch the images a priori, before the Gatsby pipeline goes in "full force" and store them in a folder that will be picked up `gatsby-source-filesystem` and `gatsby-image` and it's accompanying plugins `gatsby-transformer-sharp` and `gatsby-plugin-sharp`. More on the lifecycle apis [here](https://www.gatsbyjs.org/docs/gatsby-lifecycle-apis/)

**Comment by ArcaneTSGK** [comment]
Thanks for the responses, so I have decided that a plugin should do what we need,

The CMS is Cofoundry but it is abstracted behind a Graph QL endpoint that we have created as Middleware so that if we ever changed CMS in the future the front end is not affected as it only has to concern itself with our custom Graph QL implementation. With that said, here is where I am at.

I have a plugin which uses axios.post in the fetchData method to query the GQL endpoint and return the managed content:

```
  //Fetch Data
  let managedContentResult = fetchData({
    apiURL,
    contentQuery,
    jwtToken
  })
```

The data looks like the following (of course a lot more data is returned but to illustrate the structure)

```
{
  homePage {
    title
    thumbnail
  }
}
```
so with this managedContentResult I now need to be able to loop through and determine what is an Image to create the image nodes, and then what isn't an image to create the data nodes,

In the end I expect to end up with something like the following that would show in Graphiql and hence be able to be used on the pages etc...

```
allSiteQuery {
  edges {
     node {
       homePage {...}
     }
  }
}
```

So I have my starting point, I know where I need to be at the end, but if anyone could help out with the middle of being able to create all the nodes from this GraphQL data that is returned that would be amazing.

Thanks
